Resuse writes, initalize queues

benaadams opened the pull request:
Reuse UvWriteReq on the socket rather than recreating for each write - pooled at per connection level.

Pre initialize the circular buffers; rather than resizing as work is added; with a copy on resize.

Saves 1.12 MB per 4k requests; see https://github.com/aspnet/KestrelHttpServer/pull/363#issuecomment-159145892 for details.

Resolves #165 Pool UvWriteReq objects
Resolves #289 High Allocator DoWriteIfNeeded
Resolves #290 High Allocator SocketOutput.Write

benaadams commented:
meh

halter73 commented:
@benaadams When profiling, we see a fair amount of time spent resizing queues (though this is mostly in SocketOutput). Why did you close this PR?

benaadams commented:
I didn't feel it was worthwhile to go round the Travis repeat loop for to try to get to pass; will rebase.

benaadams commented:
Rebased and added SocketOutput

benaadams commented:
Allocation reductions for 8k requests

Before 32,023 objects; 2,032,968 bytes

![before](http://aoa.blob.core.windows.net/aspnet/preallocate-before.png)

After 12,059 objects; 947,168 bytes

![after](http://aoa.blob.core.windows.net/aspnet/preallocate-after.png)

37% of the objects than before; 46% of the bytes of before

halter73 commented:
I think @cesarbs is pretty close to fixing our travis woes. I wouldn't bother going through the travis repeat loop until that gets in. Sorry that it's been such a pain for you.

I should be able kick off reruns for all failing PRs myself once that's in.

benaadams commented:
Removed the allocs from `DoWriteIfNeeded` so down by another 212800

benaadams commented:
And from UvWriteReq.Write

benaadams commented:
Boom! Use `MemoryPool2` rather than `new byte[buffer.Count]` in write and `SocketOutput` alloc free

DamianEdwards commented:
Rebased on 7e8a405 and measured. Basically the same. I really need to profile it to see if the size we're setting them to is high enough.

```
dev (7e8a405) - 499k
363 (2a32d33) - 501k (+0.4%)
```

AspNetSmurfLab commented:
Still spending some time in resizing it seems:
![image](https://cloud.githubusercontent.com/assets/13443546/11232706/cbf53aa2-8d6a-11e5-9b15-d1fbdfc370b4.png)

benaadams commented:
@DamianEdwards that's quite a nice tool; what are you using? I'm have a hard time measuring GC with core; but 44.5% still ouch!

mattwarren commented:
@benaadams I'm pretty sure that screenshot is from [JetBrains DotTrace](https://www.google.co.uk/search?q=jetbrains+dottrace+screenshots+3.5&espv=2&biw=1257&bih=929&source=lnms&tbm=isch&sa=X&ved=0CAcQ_AUoAmoVChMIscS6qfWZyQIVxAcsCh0GdAVY&dpr=1#tbm=isch&q=jetbrains+dottrace+screenshot&imgrc=6596Vzm1wXxoAM%3A)

khellang commented:
Yep, looks like and older version of dotTrace :smile:

benaadams commented:
With 45% GC load though all bets are off on algorithmic improvements; need to get that overhead down first.

mattwarren commented:
Out of interest, which version of the GC were these tests run under? Server, Client, Background Server, etc?

benaadams commented:
@DamianEdwards  could you retest? Obv resizing the queues will add to GC pressure.

Have increased the queues from and array of 16 pointers to an array of 64 pointers - shouldn't cause much difference in allocation expense i.e. objects allocated is the same and they are reused.

AspNetSmurfLab commented:
Good results, nearly 10% improvement:
```
Running 10s test @ http://10.0.0.100:5001/plaintext
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     4.17ms    2.70ms 181.45ms   74.48%
    Req/Sec    17.22k     1.20k   29.55k    85.90%
  5518027 requests in 10.10s, 694.64MB read
Requests/sec: 546339.52
Transfer/sec:     68.78MB
```

DamianEdwards commented:
@halter73 @davidfowl @loudej pls review

davidfowl commented:
This change isn't just queue sizes it's pooling he write context and it will conflict with the socket output changes

benaadams commented:
Yes this needs revisiting post socket output

benaadams commented:
Update!

Rebased; renamed, changed to work with the SocketOutput changes. Results as follows....

Before WriteAsync bytes used 352,000 bytes

![write-async-before](http://aoa.blob.core.windows.net/aspnet/writeasync-higher.png)

After WriteAsync bytes used 888 bytes; 888 bytes? Crazy town!

![write-async-after](http://aoa.blob.core.windows.net/aspnet/writeasync-lowlowlow.png)

Before DoWriteIfNeeded 768,000 bytes

![DoWrite](http://aoa.blob.core.windows.net/aspnet/dowrite-higher.png)

After DoWriteIfNeeded doesn't use any bytes at all, so I have nothing to screenshot O_o

So instead, the full SocketOutput stack:

Before

![socket-output-before](http://aoa.blob.core.windows.net/aspnet/socketoutput-total-higher.png)

After

![socket-output-after](http://aoa.blob.core.windows.net/aspnet/socketoutput-total-lowlowlow.png)

The mind boggles... :dancer:

benaadams commented on src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs:
```diff
@@ -440,7 +481,11 @@ public void DoWriteIfNeeded()
                 }, this);
 
                 Self._head = _lockedEnd.Block;
-                Self._head.Start = _lockedEnd.Index;
```
Discovered race here; don't know if was with these changes or is in current build.

halter73 replied:
Nice catch.

If you just want to avoid the NullRefEx, you should do the following:

    var newHead = _lockedEnd.Block;
    newHead.Start = _lockedEnd.Index;

    if (Self._head != null)
    {
        Self._head = newHead;
    }

With the current change, you could still get a null ref if `_head` is set to null between the if comparison and the actual assignment to `Self._head.Start`. This makes the race a lot tighter than it was before, but it's still there.

To be safe, all modifications to _head should be made with the `_returnLock`. This fixes the race because `_head` is only set to null when the `_returnLock` is taken.

halter73 replied:
**Wait a minute!** I think this race must have been caused by a change in this PR

I was wondering how I didn't see this was a race when I originally wrote this code, and now I'm pretty sure it's because `_head` is only ever supposed to modified in the SocketOutput ctor and on the libuv thread. If that's the case, this race should be impossible.

@benaadams Do you have any idea how `_head` would be modified off of the libuv thread with this change? I'm not seeing it yet.

benaadams replied:
All I can think is the `Write` callback actually completes before the next lines are run? Was going to move the set above the `Write` but didn't feel comfortable enough with the code.

halter73 replied:
@benaadams You don't want to move the set above the write in case write throws. If that's the case, we still want to ensure that the blocks get returned to the pool eventually.

Even if the write callback gets executed inline `Self._head` should never be null after `Self._head = _lockedEnd.Block`. We already know `_lockedEnd.Block` can't be null because `LockWrite` would exit early leaving `ByteCount` at zero.

benaadams replied:
Wow, ok it is because the `Write` callback is completing before the rest of `DoWriteIfNeeded` so `OnWriteCompleted` is called before `_head` is set to `_lockedEnd.Block`

![callstack](http://aoa.blob.core.windows.net/aspnet/callstack.png)

benaadams replied:
Ahhhhh.... its the behaviour of `MockLibuv` that's causing this.

benaadams replied:
Fixed it but there would be a recursion error as `WriteAllPending` is called at the end of `OnWriteComplete` so when the stack unwinds _head could be set to an old value. Mitigating...

benaadams replied:
Added throttle fix to https://github.com/aspnet/KestrelHttpServer/pull/427

halter73 commented on src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs:
```diff
@@ -370,6 +385,32 @@ private void ReturnAllBlocks()
+            }
+        }
+
+        public void Dispose()
```
Who calls this? Is SocketOutput disposable?

benaadams replied:
No, probably should be private? (Or renamed) - `DoDisconnectIfNeeded` calls it.

halter73 replied:
I'd make it private since it doesn't implement the interface.

benaadams commented:
Bug fix in this PR moving back to `ScheduleWrite` from `WriteAllPending`; however as that may cause a regression have added https://github.com/aspnet/KestrelHttpServer/pull/427 to resolve the regression.

benaadams commented:
Rebased

benaadams commented:
Still one of the big players after recent changes; nice to see `byte[]` has dropped of the list though

![write-context](http://aoa.blob.core.windows.net/aspnet/write-context.png)

AspNetSmurfLab commented:
Here are the numbers from wrk when running the plaintext benchmark with and without this change:

__Baseline:__

	wrk -c 256 -t 32 -d 30 -s ./scripts/pipeline.lua http://10.0.0.100:5001/plaintext -- 16

	Running 30s test @ http://10.0.0.100:5001/plaintext
	  32 threads and 256 connections
	  Thread Stats   Avg      Stdev     Max   +/- Stdev
	    Latency    14.00ms   37.09ms 443.35ms   92.45%
	    Req/Sec    21.68k     2.62k  139.04k    89.94%
	  20722668 requests in 30.10s, 2.55GB read
	Requests/sec: 688462.29
	Transfer/sec:     86.67MB

	Running 30s test @ http://10.0.0.100:5001/plaintext
	  32 threads and 256 connections
	  Thread Stats   Avg      Stdev     Max   +/- Stdev
	    Latency    15.18ms   36.59ms 640.33ms   90.76%
	    Req/Sec    21.75k     2.69k  146.35k    90.92%
	  20786049 requests in 30.10s, 2.56GB read
	Requests/sec: 690577.45
	Transfer/sec:     86.93MB

	Running 30s test @ http://10.0.0.100:5001/plaintext
	  32 threads and 256 connections
	  Thread Stats   Avg      Stdev     Max   +/- Stdev
	    Latency    16.44ms   45.55ms   1.09s    92.51%
	    Req/Sec    21.65k     2.28k  185.05k    88.12%
	  20695354 requests in 30.10s, 2.54GB read
	Requests/sec: 687599.80
	Transfer/sec:     86.56MB


__After merge:__

	Running 30s test @ http://10.0.0.100:5001/plaintext
	  32 threads and 256 connections
	  Thread Stats   Avg      Stdev     Max   +/- Stdev
	    Latency     4.26ms    5.24ms 137.84ms   95.14%
	    Req/Sec    21.69k     1.11k   34.26k    73.50%
	  20739007 requests in 30.04s, 2.55GB read
	Requests/sec: 690491.69
	Transfer/sec:     86.92MB

	Running 30s test @ http://10.0.0.100:5001/plaintext
	  32 threads and 256 connections
	  Thread Stats   Avg      Stdev     Max   +/- Stdev
	    Latency     4.24ms    5.42ms 218.06ms   95.36%
	    Req/Sec    21.74k     1.07k   48.29k    72.91%
	  20789200 requests in 30.05s, 2.56GB read
	Requests/sec: 691930.78
	Transfer/sec:     87.10MB

	Running 30s test @ http://10.0.0.100:5001/plaintext
	  32 threads and 256 connections
	  Thread Stats   Avg      Stdev     Max   +/- Stdev
	    Latency     4.21ms    5.03ms 150.64ms   94.96%
	    Req/Sec    21.58k     1.03k   30.87k    69.44%
	  20637557 requests in 30.03s, 2.54GB read
	Requests/sec: 687156.69
	Transfer/sec:     86.50MB

Nothing too significant RPS-wise.

benaadams commented:
@halter73 Yeah, but look at latency...!

benaadams commented:
Rebased

benaadams commented:
I'll take the x4 latency improvement with no detriment to RPS :+1:

muratg commented:
@benaadams LogLevel.Verbose was recently renamed to LogLevel.Trace (https://github.com/aspnet/Announcements/issues/124).  Do you have the up-to-date logging packages?

khellang commented:
@muratg It fails both locally and on the CI server; https://ci.appveyor.com/project/aspnet-ci/kestrelhttpserver/build/1.0.867#L509.

Also, the Travis build seems to be stuck; https://travis-ci.org/aspnet/KestrelHttpServer/pull_requests

muratg commented:
I guess the updated packages were not pushed to the feeds yet for some reason.  cc: @pranavkm

benaadams commented:
@muratg timing? Feeds updated 20 hours ago logging change went in 19 hours ago.

halter73 commented:
We're going to need to get new numbers: https://github.com/aspnet/KestrelHttpServer/pull/458#issuecomment-162992708

benaadams commented:
Merged  #427 with this as it contains a regress fix.

halter73 commented:
@benaadams What do you think of [this idea](https://github.com/aspnet/KestrelHttpServer/commit/9c2706650912bac974cab373f79d150826a13c3f) of pooling UvWriteReqs instead of WriteContext. I started this change with the intention of continuing to pool WriteContexts, but have a shared pool per thread. I quickly realized this wouldn't be possible (without locking or a thread-safe data structure) since SocketOutput.WriteAsync is called from an arbitrary ThreadPool thread.

So at the cost of one extra allocation (that of the WriteContext), we can now have a bigger pool shared by all connections managed by another thread. I am thinking about opening a PR with this change, but since this is an alternative to pooling WriteContexts, I figured we could discuss it here first.

benaadams commented:
@halter73 Well... minimising `UvWriteReq` in all ways can only be a good thing being both a finalized object and having allocated native memory. 

However... `WriteContext` is a pretty big object; so what about pooling the `UvWriteReq` on the thread as you've done; but also pooling the `WriteContext` on the socket as it was before (so two queue references)?

halter73 commented:
I'm not sure `WriteContext` is big enough to be worth pooling. Isn't the `UvWriteReq` the main component of `WriteContext` that currently makes it big?

The `WriteContext` pool only living for the lifetime of a connection instead of the lifetime of the server also makes me disinclined to add it.

benaadams commented:
On x64 the size of `WriteContext` without the `UvWriteReq` allocation is

Component| bytes|total
----|-----|-----
MemoryPoolIterator2 _lockedStart;|-|0
    ptr _block;|8|8
    int _index;|4|12
MemoryPoolIterator2 _lockedEnd;|-|12
    ptr _block;|8|20
    int _index;|4|24
int _bufferCount;|4|28
int ByteCount;|4|32
ptr Self;|8|40
bool SocketShutdownSend;|4|44
bool SocketDisconnect;|4|48
int WriteStatus;|4|52
ptr WriteError;|8|60
ptr _writeReq;|8|68
int ShutdownSendStatus;|4|72
object _syncblk |4|76
object TypeHandle |8|84
object alignment|4|88

88 bytes; for 4k requests (write+shutdown+disconnect) where they only have a single write 1 MB

The number of writes would be higher than the number of requests as there is generally more than 1 write per request.

However, I agree with moving `UvWriteReq` to a more global pool rather than pooling with `WriteContext` as it has higher costs (finalizer+native memory)

benaadams commented:
@halter73 which would be [this on top](https://github.com/benaadams/KestrelHttpServer/commit/9ed2b85e7c36eb98c1bc81bc5f4810c9aa24ee74) of yours; look good?
